### PR TITLE
Support LLVM >=17.0.1 on Darwin and Linux

### PR DESCRIPTION
**Pull request by jcmdln**
This patch allows conditionally using the new PassBuilder on non-Windows platforms when the version of LLVM is >=17.0.1 as all other Transforms in `llvm-c` have been removed as of this version. Older supported versions of LLVM will continue using the old transforms.

Sanitizers are also now wired up for Darwin and Linux.

Ensure `/path/to/llvm17/bin` is in `PATH` or manually prepend it before building Odin:

```sh
export PATH="/path/to/llvm17/bin:$PATH"
./build_odin.sh
./odin run examples/demo/demo.odin -file -sanitize:address
```

### CODE REVIEW

**Comment by laytan** [comment]
After changing https://github.com/odin-lang/Odin/blob/cffa035c1b9d36c2639f6b7fd48b6399653244a3/src/main.cpp#L92 it also compiles on OSX

**Comment by jcmdln** [comment]
Current workarounds for using sanitizers on non-Windows platforms:

```sh
export PATH="path/to/llvm/bin:$PATH"
./build_odin.sh
./odin run examples/demo/demo.odin -file -sanitize:address -extra-linker-flags="-fsanitize=address"
```

File: build_odin.sh [hunk 1]
```diff
@@ -52,10 +52,10 @@ config_darwin() {
 		fi
 	fi
 
-	MAX_LLVM_VERSION=("14.999.999")
+	MAX_LLVM_VERSION=("17.999.999")
```
Comment (gingerBill) [inline build_odin.sh:55]: This is where things get a little complicated.

We supported LLVM 11, 12, 13, and now 17. We didn't support 14 (even if people got it working) and we definitely did not support 15 and 16. So Just a simple "max version" check is not necessarily a good idea.

Comment (gingerBill) [reply build_odin.sh:55]: But maybe this is a non-issue in practice because of the preprocessor check.

Comment (jcmdln) [reply build_odin.sh:55]: Yeah I'm curious why the shell script even needs to separately define dependencies other than to error as early as possible.

Comment (jcmdln) [reply build_odin.sh:55]:
```
$ LLVM_CONFIG=llvm-config-16 ./build_odin.sh 
...
src/llvm_backend.hpp:2:2: error: "LLVM version must be 11, 12, 13, or 17"
    2 | #error "LLVM version must be 11, 12, 13, or 17"
      |  ^
src/llvm_backend.hpp:25:10: fatal error: 'llvm-c/Transforms/AggressiveInstCombine.h' file not found
   25 | #include <llvm-c/Transforms/AggressiveInstCombine.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```
Hoisting the preprocessor check in `llvm_backend.hpp` could make this a little more clear?

Comment (laytan) [reply build_odin.sh:55]: I think a lot of people are using llvm14 at the moment and making it an error at this point should be avoided

Comment (jcmdln) [reply build_odin.sh]: I've opted to clean up `build_odin.sh` instead of relying solely on preprocessor macros.

File: src/llvm_backend.hpp [hunk 2]
```diff
@@ -15,6 +15,9 @@
 #include <llvm-c/Object.h>
 #include <llvm-c/BitWriter.h>
 #include <llvm-c/DebugInfo.h>
+#if LLVM_VERSION_MAJOR >= 17 && (LLVM_VERSION_MINOR > 0 || (LLVM_VERSION_MINOR == 0 && LLVM_VERSION_PATCH > 0))
```
Comment (gingerBill) [inline src/llvm_backend.hpp:18]: I think testing for `LLVM_VERSION_MAJOR >= 17` is all that is necessary. Checking for the minor/major distinctions is probably not a good idea.

Comment (jcmdln) [reply src/llvm_backend.hpp]: You're right, it should match the `LB_USE_NEW_PASS_SYSTEM` check below.

**Pull request merged**